### PR TITLE
Update RbConfig TRY_LINK to try to link using mock truffle c impl

### PR DESCRIPTION
**Pull request by bjfish**
While trying to `gem install pg` I ran into an issue where `mkmf` was reporting to have a function that is actually not present in `truffleruby`:
`have_func: checking for rb_str_modify_expand()... -------------------- yes`

This was due to successfully compiling the second example program it attempts below.  After updating the `TRY_LINK` to link against the included `trufflemock.bc` in this PR, this now correctly reports to not have this function:
`checking for rb_str_modify_expand()... no`

New `mkmf.log`
```
clang -o conftest /Users/bfish/Documents/truffleruby/lib/cext/ruby.bc /Users/bfish/Documents/truffleruby/lib/cext/trufflemock.bc conftest.c -I/Users/bfish/Documents/truffleruby/lib/cext -I/Users/bfish/Documents/truffleruby/lib/cext/ruby/backward -I/Users/bfish/Documents/truffleruby/lib/cext -I. -g -Werror=implicit-function-declaration -Wno-unknown-warning-option -Wno-int-conversion -Wno-int-to-pointer-cast -Wno-macro-redefined -Wno-unused-value -Wno-incompatible-pointer-types -Wno-expansion-to-defined -ferror-limit=500  -lpq   "
Undefined symbols for architecture x86_64:
  "_rb_str_modify_expand", referenced from:
      _t in conftest-c0815b.o
ld: symbol(s) not found for architecture x86_64
```

Error `mkmf.log`
```
have_func: checking for rb_str_modify_expand()... -------------------- yes

"clang -o conftest conftest.c -I/Users/bfish/Documents/truffleruby/lib/cext -I/Users/bfish/Documents/truffleruby/lib/cext/ruby/backward -I/Users/bfish/Documents/truffleruby/lib/cext -I. -g -Werror=implicit-function-declaration -Wno-unknown-warning-option -Wno-int-conversion -Wno-int-to-pointer-cast -Wno-macro-redefined -Wno-unused-value -Wno-incompatible-pointer-types -Wno-expansion-to-defined -c -emit-llvm -ferror-limit=500  -lpq   "
clang-4.0: warning: -lpq: 'linker' input unused [-Wunused-command-line-argument]
conftest.c:13:57: error: use of undeclared identifier 'rb_str_modify_expand'
int t(void) { void ((*volatile p)()); p = (void ((*)()))rb_str_modify_expand; return !p; }
                                                        ^
1 error generated.
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: /*top*/
 4: extern int t(void);
 5: int main(int argc, char **argv)
 6: {
 7:   if (argc > 1000000) {
 8:     printf("%p", &t);
 9:   }
10: 
11:   return 0;
12: }
13: int t(void) { void ((*volatile p)()); p = (void ((*)()))rb_str_modify_expand; return !p; }
/* end */

"clang -o conftest conftest.c -I/Users/bfish/Documents/truffleruby/lib/cext -I/Users/bfish/Documents/truffleruby/lib/cext/ruby/backward -I/Users/bfish/Documents/truffleruby/lib/cext -I. -g -Werror=implicit-function-declaration -Wno-unknown-warning-option -Wno-int-conversion -Wno-int-to-pointer-cast -Wno-macro-redefined -Wno-unused-value -Wno-incompatible-pointer-types -Wno-expansion-to-defined -c -emit-llvm -ferror-limit=500  -lpq   "
clang-4.0: warning: -lpq: 'linker' input unused [-Wunused-command-line-argument]
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: /*top*/
 4: extern int t(void);
 5: int main(int argc, char **argv)
 6: {
 7:   if (argc > 1000000) {
 8:     printf("%p", &t);
 9:   }
10: 
11:   return 0;
12: }
13: extern void rb_str_modify_expand();
14: int t(void) { rb_str_modify_expand(); return 0; }
/* end */

--------------------
```

### CODE REVIEW

File: src/main/c/Makefile [hunk 1]
```diff
@@ -11,7 +11,7 @@ endif
 no_cexts: clean
 	@echo "WARNING: TRUFFLERUBY_CEXT_ENABLED is set to false - not building cexts" 1>&2
 
-all: $(ROOT)/lib/cext/ruby.su $(ROOT)/lib/mri/openssl.su $(ROOT)/lib/mri/zlib.su $(ROOT)/lib/mri/syslog.su
+all: $(ROOT)/lib/cext/ruby.su $(ROOT)/lib/mri/openssl.su $(ROOT)/lib/mri/zlib.su $(ROOT)/lib/mri/syslog.su $(ROOT)/lib/cext/trufflemock.bc
 
 clean:
 	rm -f $(ROOT)/lib/cext/ruby.su
```
Comment (eregon) [inline src/main/c/Makefile]: Should also remove `ruby.bc`

File: src/main/c/trufflemock/trufflemock.c [hunk 2]
```diff
@@ -0,0 +1,309 @@
+/*
```
Comment (eregon) [inline src/main/c/trufflemock/trufflemock.c]: This file should be generated from `truffle.h`, it should be relatively straightforward with a small Ruby script.
For now, I think the re-generation can be manual. But maybe the `extconf.rb` could just re-generate every time.

File: src/main/c/trufflemock/trufflemock.c [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ *
+ */
+
+#include <stdio.h>
+#include <truffle.h>
+
+void rb_tr_mock() {
+  printf("Warning: Mock method called in trufflemock\n");
```
Comment (eregon) [inline src/main/c/trufflemock/trufflemock.c]: It's best to print errors on stderr.

File: src/main/c/Makefile (hunk 1 above)
Comment (eregon) [inline src/main/c/Makefile]: Ah, its done below, but it's better to keep them together since they are also generated by the same Makefile rule.

**Comment by eregon** [comment]
Looks good to me and it's faithful to what happens in MRI.

File: src/main/c/trufflemock/trufflemock.c (hunk 2 above)
Comment (bjfish) [inline src/main/c/trufflemock/trufflemock.c]: @eregon I've added a generate tool script and updated the `trufflemock.c` file to its output.

File: src/main/c/Makefile [hunk 4]
```diff
@@ -11,17 +11,20 @@ endif
 no_cexts: clean
 	@echo "WARNING: TRUFFLERUBY_CEXT_ENABLED is set to false - not building cexts" 1>&2
 
-all: $(ROOT)/lib/cext/ruby.su $(ROOT)/lib/mri/openssl.su $(ROOT)/lib/mri/zlib.su $(ROOT)/lib/mri/syslog.su
+all: $(ROOT)/lib/cext/trufflemock.bc $(ROOT)/lib/cext/ruby.su $(ROOT)/lib/mri/openssl.su $(ROOT)/lib/mri/zlib.su $(ROOT)/lib/mri/syslog.su
```
Comment (bjfish) [inline src/main/c/Makefile]: I had to change the order here which hopefully should get the build passing. It was failing to try to link against the non-existing `trufflemock.bc` which didn't exist yet.

File: src/main/c/Makefile (hunk 1 above)
Comment (bjfish) [inline src/main/c/Makefile]: Fixed.

File: src/main/c/trufflemock/trufflemock.c (hunk 3 above)
Comment (bjfish) [inline src/main/c/trufflemock/trufflemock.c]: Fixed.

File: src/main/c/Makefile (hunk 4 above)
Comment (eregon) [inline src/main/c/Makefile]: One nicer way to do this is to express in the dependencies below.
I think we should make `cext/Makefile:` depend on `trufflemock.bc`.

File: tool/generate-trufflemock.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+# Eclipse Public License version 1.0
+# GNU General Public License version 2
+# GNU Lesser General Public License version 2.1
+#
+
+require 'erb'
+
+types = { 'void *' => "0", 'void ' => nil, 
+    'bool ' => 'false', 'int ' => '0', 'long ' => '0', 'char ' => "'0'",
+    'float ' => '0.0', 'double ' => '0.0'}
```
Comment (eregon) [inline tool/generate-trufflemock.rb]: I think we should avoid the confusing extra space in keys here, and use `match[1].gsub(' ', '')` below.

File: tool/generate-trufflemock.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+methods = []
+
+lines = IO.readlines("lib/cext/truffle.h")
+lines.each do |l|
+   match = !l.start_with?('//') && /(.*?)truffle(.*?)\)/.match(l)
```
Comment (eregon) [inline tool/generate-trufflemock.rb]: The first capture should be `/^(.+)truffle...`, so to match from the beginning of the line. Non-greedy seems strange here.

Comment (bjfish) [inline tool/generate-trufflemock.rb]: I'm having trouble with the regex matching group `bool truffle_is_` for regex `/^(.+)truffle(.*?)\)/`

File: tool/generate-trufflemock.rb (hunk 5 above)
Comment (bjfish) [inline tool/generate-trufflemock.rb]: Fixed.

File: src/main/c/Makefile (hunk 4 above)
Comment (bjfish) [inline src/main/c/Makefile]: Fixed

File: tool/generate-trufflemock.rb (hunk 6 above)
Comment (eregon) [inline tool/generate-trufflemock.rb]: Ah, I see it matches the second `truffle`. You could use a word boundary anchor here:
```ruby
/^(.+)\btruffle(.*?)\)/
```
The non-greedy is actually fine too, but I would use `.+` not `.*` as that could be an empty capture.

Comment (eregon) [inline tool/generate-trufflemock.rb]: I changed the regexp in 93c27bd.

**Pull request merged**